File: src-wails/gallery.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"time"

	"github.com/wailsapp/wails/v2/pkg/runtime"
)

func (a *App) GetGalleryImages() ([]GalleryImage, error) {
	images, err := DBGetAllGalleryImages()
	if err != nil {
		return nil, err
	}

	for i := range images {
		if _, err := os.Stat(images[i].LocalPath); os.IsNotExist(err) {
			images[i].IsValid = false
		} else {
			images[i].IsValid = true
		}
	}

	return images, nil
}

func (a *App) GetGalleryFolders() ([]GalleryFolder, error) {
	execPath, _ := os.Executable()
	imageDir := filepath.Join(filepath.Dir(execPath), "images")

	if err := os.MkdirAll(imageDir, 0755); err != nil {
		return nil, err
	}

	entries, err := os.ReadDir(imageDir)
	if err != nil {
		return nil, err
	}

	var folders []GalleryFolder
	for _, entry := range entries {
		if entry.IsDir() {
			info, _ := entry.Info()
			folders = append(folders, GalleryFolder{
				Name:      entry.Name(),
				Path:      entry.Name(),
				CreatedAt: info.ModTime().Format(time.RFC3339),
			})
		}
	}

	return folders, nil
}

func (a *App) CreateGalleryFolder(name string) (*GalleryFolder, error) {
	execPath, _ := os.Executable()
	imageDir := filepath.Join(filepath.Dir(execPath), "images")
	folderPath := filepath.Join(imageDir, name)

	if err := os.MkdirAll(folderPath, 0755); err != nil {
		return nil, fmt.Errorf("创建文件夹失败: %w", err)
	}

	return &GalleryFolder{
		Name:      name,
		Path:      name,
		CreatedAt: time.Now().Format(time.RFC3339),
	}, nil
}

func (a *App) DeleteGalleryFolder(name string) error {
	execPath, _ := os.Executable()
	imageDir := filepath.Join(filepath.Dir(execPath), "images")
	folderPath := filepath.Join(imageDir, name)

	entries, err := os.ReadDir(folderPath)
	if err != nil {
		return fmt.Errorf("读取文件夹失败: %w", err)
	}

	if len(entries) > 0 {
		return fmt.Errorf("文件夹不为空，无法删除")
	}

	return os.Remove(folderPath)
}

func (a *App) MoveGalleryImageToFolder(imageId, folderPath string) error {
	images, err := DBGetAllGalleryImages()
	if err != nil {
		return err
	}

	var targetImage *GalleryImage
	for _, img := range images {
		if img.ID == imageId {
			targetImage = &img
			break
		}
	}

	if targetImage == nil {
		return fmt.Errorf("图片不存在")
	}

	execPath, _ := os.Executable()
	imageDir := filepath.Join(filepath.Dir(execPath), "images")

	var newPath string
	if folderPath == "" {
		newPath = filepath.Join(imageDir, filepath.Base(targetImage.LocalPath))
	} else {
		destDir := filepath.Join(imageDir, folderPath)
		if err := os.MkdirAll(destDir, 0755); err != nil {
			return fmt.Errorf("创建目标文件夹失败: %w", err)
		}
		newPath = filepath.Join(destDir, filepath.Base(targetImage.LocalPath))
	}

	if newPath != targetImage.LocalPath {
		if err := os.Rename(targetImage.LocalPath, newPath); err != nil {
			return fmt.Errorf("移动文件失败: %w", err)
		}

		_, err = db.Exec(`UPDATE wt_image SET local_path = ?, folder_path = ? WHERE id = ?`,
			newPath, folderPath, imageId)
		if err != nil {
			os.Rename(newPath, targetImage.LocalPath)
			return fmt.Errorf("更新数据库失败: %w", err)
		}
	} else {
		return DBUpdateGalleryImageFolder(imageId, folderPath)
	}

	return nil
}

func (a *App) SelectAndAddGalleryImage() (*GalleryImage, error) {
	return a.SelectAndAddGalleryImageToFolder("")
}

func (a *App) SelectAndAddGalleryImageToFolder(folderPath string) (*GalleryImage, error) {
	filePath, err := runtime.OpenFileDialog(a.ctx, runtime.OpenDialogOptions{
		Title: "选择图片",
		Filters: []runtime.FileFilter{
			{DisplayName: "图片文件", Pattern: "*.png;*.jpg;*.jpeg;*.gif;*.webp;*.bmp"},
		},
	})
	if err != nil {
		return nil, err
	}
	if filePath == "" {
		return nil, nil
	}

	return a.AddGalleryImageFromPathToFolder(filePath, folderPath)
}

func (a *App) AddGalleryImageFromPath(filePath string) (*GalleryImage, error) {
	return a.AddGalleryImageFromPathToFolder(filePath, "")
}

func (a *App) AddGalleryImageFromPathToFolder(filePath string, folderPath string) (*GalleryImage, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("无法打开文件: %w", err)
	}
	defer file.Close()

	hasher := sha256.New()
	if _, err := io.Copy(hasher, file); err != nil {
		return nil, fmt.Errorf("计算哈希失败: %w", err)
	}
	hash := hex.EncodeToString(hasher.Sum(nil))

	existing, err := DBGetGalleryImageByHash(hash)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		existing.IsValid = true
		return existing, nil
	}

	fileInfo, err := os.Stat(filePath)
	if err != nil {
		return nil, fmt.Errorf("获取文件信息失败: %w", err)
	}

	execPath, _ := os.Executable()
	imageDir := filepath.Join(filepath.Dir(execPath), "images")
	if folderPath != "" {
		imageDir = filepath.Join(imageDir, folderPath)
	}
	if err := os.MkdirAll(imageDir, 0755); err != nil {
		return nil, fmt.Errorf("创建图片目录失败: %w", err)
	}

	ext := filepath.Ext(filePath)
	newFileName := hash[:16] + ext
	destPath := filepath.Join(imageDir, newFileName)

	file.Seek(0, 0)
	destFile, err := os.Create(destPath)
	if err != nil {
		return nil, fmt.Errorf("创建目标文件失败: %w", err)
	}
	defer destFile.Close()

	if _, err := io.Copy(destFile, file); err != nil {
		return nil, fmt.Errorf("复制文件失败: %w", err)
	}

	img := &GalleryImage{
		ID:         fmt.Sprintf("img_%d", time.Now().UnixNano()),
		Hash:       hash,
		LocalPath:  destPath,
		FileName:   filepath.Base(filePath),
		FileSize:   fileInfo.Size(),
		CreatedAt:  time.Now().Format(time.RFC3339),
		FolderPath: folderPath,
		IsValid:    true,
	}

	if err := DBAddGalleryImage(img); err != nil {
		os.Remove(destPath)
		return nil, fmt.Errorf("保存图片记录失败: %w", err)
	}

	return img, nil
}

func (a *App) AddGalleryImageFromBase64(base64Data string, fileName string) (*GalleryImage, error) {
	return a.AddGalleryImageFromBase64ToFolder(base64Data, fileName, "")
}

func (a *App) AddGalleryImageFromBase64ToFolder(base64Data string, fileName string, folderPath string) (*GalleryImage, error) {
	idx := 0
	for i := 0; i < len(base64Data); i++ {
		if base64Data[i] == ',' {
			idx = i + 1
			break
		}
	}
	if idx == 0 {
		idx = 0
	}
	data, err := base64Decode(base64Data[idx:])
	if err != nil {
		return nil, fmt.Errorf("base64解码失败: %w", err)
	}
	return a.AddGalleryImageFromBytesToFolder(data, fileName, folderPath)
}

func (a *App) AddGalleryImageFromBytes(data []byte, fileName string) (*GalleryImage, error) {
	return a.AddGalleryImageFromBytesToFolder(data, fileName, "")
}

func (a *App) AddGalleryImageFromBytesToFolder(data []byte, fileName string, folderPath string) (*GalleryImage, error) {
	hasher := sha256.New()
	hasher.Write(data)
	hash := hex.EncodeToString(hasher.Sum(nil))

	existing, err := DBGetGalleryImageByHash(hash)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		existing.IsValid = true
		return existing, nil
	}

	execPath, _ := os.Executable()
	imageDir := filepath.Join(filepath.Dir(execPath), "images")
	if folderPath != "" {
		imageDir = filepath.Join(imageDir, folderPath)
	}
	if err := os.MkdirAll(imageDir, 0755); err != nil {
		return nil, fmt.Errorf("创建图片目录失败: %w", err)
	}

	ext := filepath.Ext(fileName)
	if ext == "" {
		ext = ".png"
	}
	newFileName := hash[:16] + ext
	destPath := filepath.Join(imageDir, newFileName)

	if err := os.WriteFile(destPath, data, 0644); err != nil {
		return nil, fmt.Errorf("保存图片失败: %w", err)
	}

	img := &GalleryImage{
		ID:         fmt.Sprintf("img_%d", time.Now().UnixNano()),
		Hash:       hash,
		LocalPath:  destPath,
		FileName:   fileName,
		FileSize:   int64(len(data)),
		CreatedAt:  time.Now().Format(time.RFC3339),
		FolderPath: folderPath,
		IsValid:    true,
	}

	if err := DBAddGalleryImage(img); err != nil {
		os.Remove(destPath)
		return nil, fmt.Errorf("保存图片记录失败: %w", err)
	}

	return img, nil
}

func (a *App) UpdateGalleryImageURL(id, remoteURL string) error {
	return DBUpdateGalleryImageURL(id, remoteURL)
}

func (a *App) RenameGalleryImage(id, newName string) error {
	return DBUpdateGalleryImageName(id, newName)
}

func (a *App) RenameGalleryFolder(oldName, newName string) error {
	execPath, _ := os.Executable()
	imageDir := filepath.Join(filepath.Dir(execPath), "images")
	oldPath := filepath.Join(imageDir, oldName)
	newPath := filepath.Join(imageDir, newName)

	if err := os.Rename(oldPath, newPath); err != nil {
		return fmt.Errorf("重命名文件夹失败: %w", err)
	}

	if err := DBRenameFolderImages(oldName, newName); err != nil {
		os.Rename(newPath, oldPath)
		return fmt.Errorf("更新数据库失败: %w", err)
	}

	return nil
}

func (a *App) GetFolderImages(folderPath string) ([]GalleryImage, error) {
	images, err := DBGetAllGalleryImages()
	if err != nil {
		return nil, err
	}

	var result []GalleryImage
	for _, img := range images {
		if img.FolderPath == folderPath {
			if _, err := os.Stat(img.LocalPath); os.IsNotExist(err) {
				img.IsValid = false
			} else {
				img.IsValid = true
			}
			result = append(result, img)
		}
	}

	return result, nil
}

func (a *App) DeleteGalleryImage(id string) error {
	images, err := DBGetAllGalleryImages()
	if err != nil {
		return err
	}

	var targetImage *GalleryImage
	for _, img := range images {
		if img.ID == id {
			targetImage = &img
			break
		}
	}

	if targetImage == nil {
		return fmt.Errorf("图片不存在")
	}

	if err := DBDeleteGalleryImage(id); err != nil {
		return err
	}

	if targetImage.LocalPath != "" {
		os.Remove(targetImage.LocalPath)
	}

	return nil
}

func (a *App) DeleteGalleryImages(ids []string) error {
	for _, id := range ids {
		if err := a.DeleteGalleryImage(id); err != nil {
			return err
		}
	}
	return nil
}

func (a *App) GetGalleryImageLocalURL(id string) (string, error) {
	images, err := DBGetAllGalleryImages()
	if err != nil {
		return "", err
	}

	for _, img := range images {
		if img.ID == id {
			return img.LocalPath, nil
		}
	}
	return "", fmt.Errorf("图片不存在")
}

func (a *App) ReadGalleryImageAsBase64(id string) (string, error) {
	images, err := DBGetAllGalleryImages()
	if err != nil {
		return "", err
	}

	var targetImage *GalleryImage
	for _, img := range images {
		if img.ID == id {
			targetImage = &img
			break
		}
	}

	if targetImage == nil {
		return "", fmt.Errorf("图片不存在")
	}

	data, err := os.ReadFile(targetImage.LocalPath)
	if err != nil {
		return "", fmt.Errorf("读取图片失败: %w", err)
	}

	ext := filepath.Ext(targetImage.LocalPath)
	mimeType := "image/png"
	switch ext {
	case ".jpg", ".jpeg":
		mimeType = "image/jpeg"
	case ".gif":
		mimeType = "image/gif"
	case ".webp":
		mimeType = "image/webp"
	case ".bmp":
		mimeType = "image/bmp"
	}

	base64Data := "data:" + mimeType + ";base64," + base64Encode(data)
	return base64Data, nil
}

func base64Encode(data []byte) string {
	const base64Chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
	result := make([]byte, 0, (len(data)+2)/3*4)
	for i := 0; i < len(data); i += 3 {
		var n uint32
		remaining := len(data) - i
		if remaining >= 3 {
			n = uint32(data[i])<<16 | uint32(data[i+1])<<8 | uint32(data[i+2])
			result = append(result, base64Chars[n>>18&0x3F], base64Chars[n>>12&0x3F], base64Chars[n>>6&0x3F], base64Chars[n&0x3F])
		} else if remaining == 2 {
			n = uint32(data[i])<<16 | uint32(data[i+1])<<8
			result = append(result, base64Chars[n>>18&0x3F], base64Chars[n>>12&0x3F], base64Chars[n>>6&0x3F], '=')
		} else {
			n = uint32(data[i]) << 16
			result = append(result, base64Chars[n>>18&0x3F], base64Chars[n>>12&0x3F], '=', '=')
		}
	}
	return string(result)
}

func base64Decode(s string) ([]byte, error) {
	const base64Chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
	decodeMap := make(map[byte]byte)
	for i := 0; i < len(base64Chars); i++ {
		decodeMap[base64Chars[i]] = byte(i)
	}
	padCount := 0
	for i := len(s) - 1; i >= 0 && s[i] == '='; i-- {
		padCount++
	}
	s = s[:len(s)-padCount]
	result := make([]byte, 0, len(s)*3/4)
	for i := 0; i < len(s); i += 4 {
		var n uint32
		count := 4
		if i+4 > len(s) {
			count = len(s) - i
		}
		for j := 0; j < count; j++ {
			n = n<<6 | uint32(decodeMap[s[i+j]])
		}
		for j := count; j < 4; j++ {
			n = n << 6
		}
		if count >= 2 {
			result = append(result, byte(n>>16))
		}
		if count >= 3 {
			result = append(result, byte(n>>8))
		}
		if count >= 4 {
			result = append(result, byte(n))
		}
	}
	return result, nil
}

File: src-tauri/src/commands/gallery.rs
use crate::database;
use crate::types::{GalleryFolder, GalleryImage};
use chrono::Utc;
use sha2::{Sha256, Digest};
use std::fs;
use std::path::PathBuf;
use tauri::Manager;

fn get_gallery_dir(app: &tauri::AppHandle) -> Result<PathBuf, String> {
    let app_data = app.path().app_data_dir().map_err(|e| e.to_string())?;
    let gallery_dir = app_data.join("gallery");
    fs::create_dir_all(&gallery_dir).map_err(|e| e.to_string())?;
    Ok(gallery_dir)
}

#[tauri::command]
pub fn get_gallery_images() -> Result<Vec<GalleryImage>, String> {
    database::with_db_log("get_gallery_images", |conn| {
        let mut stmt = conn.prepare(
            "SELECT id, hash, local_path, remote_url, file_name, file_size, created_at, folder_path FROM wt_image ORDER BY created_at DESC"
        )?;
        let images = stmt.query_map([], |row| {
            Ok(GalleryImage {
                id: row.get(0)?,
                hash: row.get(1)?,
                local_path: row.get(2)?,
                remote_url: row.get::<_, Option<String>>(3)?.unwrap_or_default(),
                file_name: row.get::<_, Option<String>>(4)?.unwrap_or_default(),
                file_size: row.get::<_, Option<i64>>(5)?.unwrap_or(0),
                created_at: row.get(6)?,
                folder_path: row.get::<_, Option<String>>(7)?.unwrap_or_default(),
            })
        })?;
        images.collect::<Result<Vec<_>, _>>()
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_gallery_folders() -> Result<Vec<GalleryFolder>, String> {
    database::with_db_log("get_gallery_folders", |conn| {
        let mut stmt = conn.prepare(
            "SELECT folder_path, COUNT(*) as count FROM wt_image WHERE folder_path != '' GROUP BY folder_path"
        )?;
        let folders = stmt.query_map([], |row| {
            let path: String = row.get(0)?;
            Ok(GalleryFolder {
                name: path.clone(),
                path,
                image_count: row.get(1)?,
            })
        })?;
        folders.collect::<Result<Vec<_>, _>>()
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn create_gallery_folder(name: String) -> Result<GalleryFolder, String> {
    Ok(GalleryFolder {
        name: name.clone(),
        path: name,
        image_count: 0,
    })
}

#[tauri::command]
pub fn delete_gallery_folder(name: String) -> Result<(), String> {
    database::with_db_log(&format!("delete_gallery_folder: {}", name), |conn| {
        conn.execute(
            "UPDATE wt_image SET folder_path = '' WHERE folder_path = ?",
            [&name],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_gallery_image(app: tauri::AppHandle, id: String) -> Result<(), String> {
    let local_path: String = database::with_db_log(&format!("get_image_path: {}", id), |conn| {
        conn.query_row(
            "SELECT local_path FROM wt_image WHERE id = ?",
            [&id],
            |row| row.get(0),
        )
    }).map_err(|e| e.to_string())?;
    let gallery_dir = get_gallery_dir(&app)?;
    let file_path = gallery_dir.join(&local_path);
    if file_path.exists() {
        fs::remove_file(file_path).map_err(|e| e.to_string())?;
    }
    database::with_db_log(&format!("delete_gallery_image: {}", id), |conn| {
        conn.execute("DELETE FROM wt_image WHERE id = ?", [&id])?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_gallery_images(app: tauri::AppHandle, ids: Vec<String>) -> Result<(), String> {
    for id in ids {
        delete_gallery_image(app.clone(), id)?;
    }
    Ok(())
}

#[tauri::command]
pub fn move_gallery_image_to_folder(image_id: String, folder_path: String) -> Result<(), String> {
    database::with_db_log(&format!("move_gallery_image: {}", image_id), |conn| {
        conn.execute(
            "UPDATE wt_image SET folder_path = ? WHERE id = ?",
            [&folder_path, &image_id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn read_gallery_image_as_base64(app: tauri::AppHandle, id: String) -> Result<String, String> {
    let local_path: String = database::with_db_log(&format!("get_image_path: {}", id), |conn| {
        conn.query_row(
            "SELECT local_path FROM wt_image WHERE id = ?",
            [&id],
            |row| row.get(0),
        )
    }).map_err(|e| e.to_string())?;
    let gallery_dir = get_gallery_dir(&app)?;
    let file_path = gallery_dir.join(&local_path);
    let data = fs::read(&file_path).map_err(|e| e.to_string())?;
    Ok(base64::Engine::encode(&base64::engine::general_purpose::STANDARD, &data))
}

#[tauri::command]
pub async fn select_and_add_gallery_image(app: tauri::AppHandle) -> Result<GalleryImage, String> {
    use tauri_plugin_dialog::DialogExt;
    let file_path = app.dialog()
        .file()
        .add_filter("Images", &["png", "jpg", "jpeg", "gif", "webp"])
        .blocking_pick_file()
        .ok_or("No file selected")?;
    let path_str = file_path.as_path().map(|p| p.to_string_lossy().to_string()).ok_or("Invalid file path")?;
    add_gallery_image_from_path(app, path_str, String::new()).await
}

#[tauri::command]
pub async fn select_and_add_to_folder(app: tauri::AppHandle, folder_path: String) -> Result<GalleryImage, String> {
    use tauri_plugin_dialog::DialogExt;
    let file_path = app.dialog()
        .file()
        .add_filter("Images", &["png", "jpg", "jpeg", "gif", "webp"])
        .blocking_pick_file()
        .ok_or("No file selected")?;
    let path_str = file_path.as_path().map(|p| p.to_string_lossy().to_string()).ok_or("Invalid file path")?;
    add_gallery_image_from_path(app, path_str, folder_path).await
}

async fn add_gallery_image_from_path(app: tauri::AppHandle, path: String, folder_path: String) -> Result<GalleryImage, String> {
    let data = fs::read(&path).map_err(|e| e.to_string())?;
    let file_name = PathBuf::from(&path)
        .file_name()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_default();
    add_image_internal(app, data, file_name, folder_path)
}

#[tauri::command]
pub fn add_gallery_image_from_base64(app: tauri::AppHandle, base64_data: String, file_name: String) -> Result<GalleryImage, String> {
    let data = base64::Engine::decode(&base64::engine::general_purpose::STANDARD, &base64_data)
        .map_err(|e| e.to_string())?;
    add_image_internal(app, data, file_name, String::new())
}

#[tauri::command]
pub fn add_gallery_image_from_base64_to_folder(app: tauri::AppHandle, base64_data: String, file_name: String, folder_path: String) -> Result<GalleryImage, String> {
    let data = base64::Engine::decode(&base64::engine::general_purpose::STANDARD, &base64_data)
        .map_err(|e| e.to_string())?;
    add_image_internal(app, data, file_name, folder_path)
}

fn add_image_internal(app: tauri::AppHandle, data: Vec<u8>, file_name: String, folder_path: String) -> Result<GalleryImage, String> {
    let mut hasher = Sha256::new();
    hasher.update(&data);
    let hash = format!("{:x}", hasher.finalize());
    let existing: Option<String> = database::with_db_log("check_image_exists", |conn| {
        Ok(conn.query_row(
            "SELECT id FROM wt_image WHERE hash = ?",
            [&hash],
            |row| row.get(0),
        ).ok())
    }).map_err(|e: rusqlite::Error| e.to_string())?;
    if let Some(id) = existing {
        return Err(format!("Image already exists with id: {}", id));
    }
    let id = uuid::Uuid::new_v4().to_string();
    let ext = PathBuf::from(&file_name)
        .extension()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_else(|| "png".to_string());
    let local_path = format!("{}.{}", id, ext);
    let gallery_dir = get_gallery_dir(&app)?;
    let full_path = gallery_dir.join(&local_path);
    fs::write(&full_path, &data).map_err(|e| e.to_string())?;
    let now = Utc::now().format("%Y-%m-%d %H:%M:%S").to_string();
    let file_size = data.len() as i64;
    database::with_db_log(&format!("add_gallery_image: {}", id), |conn| {
        conn.execute(
            "INSERT INTO wt_image (id, hash, local_path, file_name, file_size, created_at, folder_path) VALUES (?, ?, ?, ?, ?, ?, ?)",
            rusqlite::params![id, hash, local_path, file_name, file_size, now, folder_path],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())?;
    Ok(GalleryImage {
        id,
        hash,
        local_path,
        remote_url: String::new(),
        file_name,
        file_size,
        created_at: now,
        folder_path,
    })
}

#[tauri::command]
pub fn rename_gallery_image(id: String, new_name: String) -> Result<(), String> {
    database::with_db_log(&format!("rename_gallery_image: {}", id), |conn| {
        conn.execute(
            "UPDATE wt_image SET file_name = ? WHERE id = ?",
            [&new_name, &id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn rename_gallery_folder(old_name: String, new_name: String) -> Result<(), String> {
    database::with_db_log(&format!("rename_gallery_folder: {}", old_name), |conn| {
        conn.execute(
            "UPDATE wt_image SET folder_path = ? WHERE folder_path = ?",
            [&new_name, &old_name],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_gallery_image_url(id: String, url: String) -> Result<(), String> {
    database::with_db_log(&format!("update_gallery_image_url: {}", id), |conn| {
        conn.execute(
            "UPDATE wt_image SET remote_url = ? WHERE id = ?",
            [&url, &id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}
